DNA final

alexdzn opened the pull request.

alexdzn committed:
FINAL

alexdzn committed:
FINAL

little bug fixes

Wenomer commented on hexlet_dna/hexlet_dna/Cell+Mutator.m:
```diff
@@ -0,0 +1,39 @@
+        
+        //заменяем нуклеотид по индексу indexToReplace
+        int nucleotideIndex = arc4random() % 4;
+        [self.dna replaceObjectAtIndex:indexToReplace withObject:[self.nucleotides objectAtIndex:nucleotideIndex]];
```
Вы забыли проверить что именно **заменяете** нуклеотид. У вас возможна ситуация когда, например, "С" поменяется на "С" , что по сути и не замена вовсе)

Wenomer commented on hexlet_dna/hexlet_dna/Cell+Mutator.m:
```diff
@@ -0,0 +1,39 @@
+@implementation Cell (Mutator)
+
+- (void) mutate:(int)percent {
+    if(percent > 100) percent = 100;
```
а если меньше 0?

Wenomer commented on hexlet_dna/hexlet_dna/Cell+Mutator.m:
```diff
@@ -0,0 +1,39 @@
+        
+        replaced++;
+        
+    } while (replaced < percent);
```
вы считаете что 1% это одна ячейка, в данном случае это так, а если в клетке будет ячеек != 100?

alexdzn committed:
FINAL fixes

- правильные проценты ( 1% != 1 нуклеотиду )
- проверка корректности передаваемого %
- правильная замена нуклеотида. перед заменой происходит сравнение
старого и нового значений чтобы избежать замены на тот же символ/объект

alexdzn committed:
FINAL

оптимизация производительности

tarpaha commented on hexlet_dna/hexlet_dna/Cell+Mutator.m:
```diff
@@ -0,0 +1,49 @@
+- (void) mutate:(int)percent {
+    
+    //проверка допустимыx значений
+    if(percent < 1 || percent > 100) {
```
percent == 0 вполне нормальная ситуация и, как мне кажется, должна нормально обрабатываться.

alexdzn replied:
Не думаю что есть смысл заменять 0%.

tarpaha commented on hexlet_dna/hexlet_dna/Cell.m:
```diff
@@ -0,0 +1,43 @@
+        for(int i = 0; i < _dnaLength; i++){
+            //заполняем ДНК случайными нуклеотидами
+            nucleotideIndex = arc4random() % 4;
+            [_dna insertObject:[_nucleotides objectAtIndex:nucleotideIndex] atIndex:i];
```
Заполнение пустого массива лучше делать через addObject

alexdzn replied:
Полностью согласен.
